File: 100/145 - Binary Tree Postorder Traversal.java
import java.util.ArrayList;
import java.util.Stack;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public List<Integer> postorderTraversal(TreeNode root) {
        TreeNode curr, prev = null;
        List<Integer> ret = new ArrayList<Integer>();
        Stack<TreeNode> stack = new Stack<TreeNode>();
        
        if (root == null)
            return ret;
        stack.push(root);
        while (! stack.empty()) {
            curr = stack.peek();
            if (prev == null || prev.left == curr || prev.right == curr) {
                if (curr.left != null)
                    stack.push(curr.left);
                else if (curr.right != null)
                    stack.push(curr.right);
                else
                    ret.add(stack.pop().val);
            } else if (prev == curr.left && curr.right != null)
                stack.push(curr.right);
            else
                ret.add(stack.pop().val);
            prev = curr;
        }
        
        return ret;
    }
}

File: csharp/145.binary-tree-postorder-traversal.cs
/*
 * @lc app=leetcode id=145 lang=csharp
 *
 * [145] Binary Tree Postorder Traversal
 */
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */

using System.Collections.Generic;

public class Solution {
    public IList<int> PostorderTraversal(TreeNode root) {
        var ret = new List<int>();
        if (root == null) {
            return ret;
        }
        var stack = new List<TreeNode>();
        stack.Add(root);
        TreeNode prev = root;
        while (stack.Count != 0) {
            TreeNode top = stack[stack.Count - 1];
            if ((top.left == null && top.right == null) || prev == top.left || prev == top.right) {
                ret.Add(top.val);
                stack.RemoveAt(stack.Count - 1);
                prev = top;
            } else {
                
                if (top.right != null) {
                    stack.Add(top.right);
                }
                if (top.left != null) {
                    stack.Add(top.left);
                }
            }
        }

        return ret;
    }
}

